File: frontend/src/types/bay.ts
export type VehicleSize = 'SMALL'|'MEDIUM'|'LARGE';
export type VehicleType = 'VAN_S'|'TRUCK_M'|'TRUCK_L';
export type BayType = 'GOODS_IN'|'GOODS_OUT'|'DYNAMIC'|'PARKING';
export type BayStatus = 'EMPTY'|'RESERVED'|'VEHICLE_PRESENT'|'MAINTENANCE';

export interface Bay {
  id: string;
  name: string;
  // Scope bay to a warehouse
  warehouse_id: string;
  type: BayType;
  dynamicMode?: 'GOODS_IN'|'GOODS_OUT';
  capacity: number;
  vehicleCompat: VehicleSize[];
  status: BayStatus;
  reserved_for?: { ref: string; direction: 'IN'|'OUT'; eta?: string };
  vehicle?: { reg: string; type: VehicleType; carrier?: string; vendor?: string };
  operation?: 'UNLOADING'|'LOADING';
  progressPct?: number;
  utilizationPct?: number;
  created_at: string;
  updated_at: string;
}

export interface BayKpis {
  total: number;
  occupied: number;
  reserved: number;
  maintenance: number;
  empty: number;
  utilization: number; // avg across bays
  averageTurnaroundMin: number;
  idleRatePct: number;
  efficiencyPct: number;
}

File: backend/app/schemas/bay.py
from __future__ import annotations
from typing import Literal, Optional, List
from pydantic import BaseModel


VehicleSize = Literal['SMALL', 'MEDIUM', 'LARGE']
VehicleType = Literal['VAN_S', 'TRUCK_M', 'TRUCK_L']
BayType = Literal['GOODS_IN', 'GOODS_OUT', 'DYNAMIC', 'PARKING']
BayStatus = Literal['EMPTY', 'RESERVED', 'VEHICLE_PRESENT', 'MAINTENANCE']


class BayReservedFor(BaseModel):
    ref: str
    direction: Literal['IN', 'OUT']
    eta: Optional[str] = None


class BayVehicle(BaseModel):
    reg: str
    type: VehicleType
    carrier: Optional[str] = None
    vendor: Optional[str] = None


class Bay(BaseModel):
    id: str
    name: str
    warehouse_id: str
    type: BayType
    dynamicMode: Optional[Literal['GOODS_IN', 'GOODS_OUT']] = None
    capacity: int
    vehicleCompat: List[VehicleSize]
    status: BayStatus
    reserved_for: Optional[BayReservedFor] = None
    vehicle: Optional[BayVehicle] = None
    operation: Optional[Literal['UNLOADING', 'LOADING']] = None
    progressPct: Optional[int] = None
    utilizationPct: Optional[int] = None
    created_at: str
    updated_at: str


class BayCreate(BaseModel):
    id: str
    name: str
    warehouse_id: str
    type: BayType
    dynamicMode: Optional[Literal['GOODS_IN', 'GOODS_OUT']] = None
    capacity: int
    vehicleCompat: List[VehicleSize]
    status: BayStatus


class BayUpdate(BaseModel):
    name: Optional[str] = None
    type: Optional[BayType] = None
    dynamicMode: Optional[Literal['GOODS_IN', 'GOODS_OUT']] = None
    capacity: Optional[int] = None
    vehicleCompat: Optional[List[VehicleSize]] = None
    status: Optional[BayStatus] = None
    reserved_for: Optional[BayReservedFor] = None
    vehicle: Optional[BayVehicle] = None
    operation: Optional[Literal['UNLOADING', 'LOADING']] = None
    progressPct: Optional[int] = None


class BayAssignRequest(BaseModel):
    direction: Literal['IN', 'OUT']
    ref: str
    eta: Optional[str] = None
    vehicle: Optional[BayVehicle] = None
    operation: Optional[Literal['UNLOADING', 'LOADING']] = None


class BayKpis(BaseModel):
    total: int
    occupied: int
    reserved: int
    maintenance: int
    empty: int
    utilization: int
    averageTurnaroundMin: int
    idleRatePct: int
    efficiencyPct: int
